### ISSUE TITLE
"switch/case" component bundles scripts for all cases

### ISSUE THREAD

**Issue by y-nk** [withastro/astro#8223]
### Astro info

```block
Astro version            v2.10.14
Package manager          npm
Platform                 linux
Architecture             x64
Adapter                  Couldn't determine.
Integrations             @astrojs/react
```

### What browser are you using?

stackblitz

### Usecase

I'm making a `Map<string, AstroComponent>` which i use to switch between multiple implementations of a single component based on an enum, something like:

```ts
type Carousel = {
  viewType?: 'standard' | 'not-animated' | 'infinite-loop'
}
```

...which ends up switching between `<StandardView />`, `<StaticView />` and `<LoopView />` components.

### Describe the Bug

I have found that when doing something such as:

```tsx
---
import View1 from './View1.astro'
import View2 from './View2.astro'
import View3 from './View3.astro'

export type Props = {
  type?: 'view1' | 'view2' | 'view3'
}

const View = (
  Astro.props.type === 'view3' ? View3 :
  Astro.props.type === 'view2' ? View2 :
  View1
)
---

<View><slot /></View>
``` 

...the `<script>` tags of `View1` `View2` and `View3` are inserted in the page, even tho only the html corresponding to the switch/case is actually rendered.

### Non working workarounds

I have tried different syntax as workaround but they don't work.

#### In template switch/case

```tsx
---
import View1 from './View1.astro'
import View2 from './View2.astro'
import View3 from './View3.astro'

export type Props = {
  type?: 'view1' | 'view2' | 'view3'
}
---

{
  Astro.props.type === 'view3' ? (
    <View3><slot /></View3>
  ) :
  Astro.props.type === 'view2' ? (
    <View2><slot /></View2>
  ) : (
    <View1><slot /></View1>
  )
}
```
→ still bundle the js of the 3 versions

#### Dynamic import

```tsx
---
export type Props = {
  type?: 'view1' | 'view2' | 'view3'
}

const mod = (
  Astro.props.type === 'view3' ? 'View3' :
  Astro.props.type === 'view2' ? 'View2' :
  'View1'
)

const { default: View } = await import (
  `./${mod}.astro`
)
---

<View><slot /></View>
```

#### bundle it all

```tsx
---
export type Props = {
  type?: 'view1' | 'view2' | 'view3'
}
---

<div data-type={Astro.props.type}><slot /></div>

<script>
  import view1 from './view1.ts'
  import view2 from './view2.ts'
  import view3 from './view3.ts'

  const behaviors = { view1, view2, view3 }

  document.querySelectorAll<HTMLElement>(div['data-type'])
    .forEach(view => {
      const type = view.getAttribute('data-type')!
      behaviors[type](view)
    })
</script>
```

→ bundles it all but at least execution is regulated from userland code.
→ DOM of different view1/view2/view3 cannot be separated, or must be in a ternary

### What's the expected result?

We should only have the <script> tags for the matching component.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-pr1dus

### Participation

- [x] I am willing to submit a pull request for this issue.

**Comment by y-nk** [comment]
I'm realizing this is quite a big issue since it could have wider consequences for everyone using an external CMS out there. 

Usually, CMS deliver page content as an array of blocks, each block having a type ; then the typical idea is to: `

```tsx
{
  blocks.map(({ type, ...config }) => {
    const Block = getBlockByType(type) // ← a function doing the switch/case work
    return <Block {...config} />
  })
}
```

If the bug above extends to this usecase ↑ then all the blocks script tags would be bundled in every page even tho they don't have to.

**Comment by bluwy** [comment]
Currently our script tag adding heuristic is via `import` and not runtime render, so if you have imported that component, it gets into the page automatically. I agree though that this isn't great and has an umbrella of other issues like https://github.com/withastro/astro/issues/7857.

However, from the implementation pov this was an intentional caveat so this isn't exactly a bug. It would be worth creating  a discussion at https://github.com/withastro/roadmap instead so we can plan to (perhaps) switch to rendering scripts through runtime rendering instead of static import analysis.

**Comment by y-nk** [comment]
@bluwy thanks a lot for your answer. I thought it'd be worth filing as a bug but I understand that this is an intentional caveat. However it is not exactly as you described.

> Currently our script tag adding heuristic is via import and not runtime render, so if you have imported that component, it gets into the page automatically.

I found by testing that this is not 100% truth. The component must be imported _and referenced at least once_ in either of frontmatter or the template for its script tag to be included in the page.

For instance, doing this:

```
---
import View1 from '../components/View1.astro'
import View2 from '../components/View2.astro'
import View3 from '../components/View3.astro'

import Button from '../components/Button.astro'
---

<html>
  <body>
    <View1>
      <Button />
    </View1>
  </body>
</html>
```

does NOT include the `<script>` tags of `<View2 />` and `<View3 />` but this:

```
---
import View1 from '../components/View1.astro'
import View2 from '../components/View2.astro'
import View3 from '../components/View3.astro'

import Button from '../components/Button.astro'

console.log({ View2, View3 })
---

<html>
  <body>
    <View1>
      <Button />
    </View1>
</html>
```

does, which even "intentional caveat", strikes me as "surprising" behavior.

**Comment by bluwy** [comment]
It's because the unused components get treeshaken away before we can analyse with, which I think is part of the static analysis caveat. We do try to not include scripts whenever possible, so the example where it's unused and we know it, we would not include the script tag.

**Comment by y-nk** [comment]
Thanks for the explanation, it makes sense to me.

According to you this should be rather filed as a "possible improvement" rather than a bug tho? 

I mean, it is expected that a single `console.log(Component)` could end up changing the page served and its behavior? Is there any good usecase to this?

**Comment by bluwy** [comment]
Yeah I think it's better to file this as a feature/improvement in the roadmap repo. We usually work on features tracked there instead.

> I mean, it is expected that a single `console.log(Component)` could end up changing the page served and its behavior? Is there any good usecase to this?

There's not a lot of reason, but it's static analysis seeing it as a "side-effect". If you `console.log` something it can't remove it because it's unexpected the logs aren't logged later.

**Comment by y-nk** [comment]
i see. i'll file a discussion in the roadmap but the difference of treatment between html and script is rather unsettling ; and the fact there's no way to branch out from this as well...

The proposed workaround on discord does not scale well neither:

```
<script>
if (document.querySelector('component-tag')) await import('./component-script.js');
</script>
```

...because of how block based CMS work. The current website i'm building we have ~80 components which are loaded based on their block type (as the example above ↑), which would mean 80 script tags in each page, containing either a delegate loader statement or the full code. Both way it's 80 script tags in each page.

**Comment by y-nk** [comment]
@lilnasy @bluwy 

closing in favor of: https://github.com/withastro/roadmap/discussions/664
